{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After loggin into the server then paste the follow into your terminal to set paths\n",
    "\n",
    "DATA=/ricco/data/PhDCourse/summer2019/ngs/\n",
    "SAMfile=$DATA/id.sam.gz\n",
    "FastQNA19238=$DATA/NA19238.fastq.gz\n",
    "CHR21=$DATA/chr21.fa.gz\n",
    "HG19=$DATA/hg19.fa.gz\n",
    "COPY=$DATA/preComp/\n",
    "\n",
    "# programs\n",
    "SAMTOOLS=samtools\n",
    "BWA=bwa\n",
    "PICARD=$DATA/picard.jar\n",
    "ANGSD=angsd\n",
    "BCFTOOL=bcftools\n",
    "\n",
    "Mapping one reduced genome (single end)\n",
    "\n",
    "In this exercise you will align a fastq file using bwa and generate a SAM file.\n",
    "\n",
    "Due to the computational time we have created a reduced genome from one of the individuals from the 1000genomes pilot project. The individual, NA19238, has been sequenced using solexa/Illumina. For this excerice we have created a reduced reference genome consisting only of chromosome 21 (the smallest human chromosome) and also reduced the sequencing data to reads that will likely map to chromosome 21.\n",
    "\n",
    "    The fastQ file 'NA19238.fastq.gz' has variable name $FastQNA19238 which is the sample we want to align and the chr21.fa.gz ($CHR21) reference genome that we want to align against.\n",
    "\n",
    "Viewing the input files\n",
    "\n",
    "    view the fastq file (NA19238.fastq) using the less command and identity the reads and quality scores (ignore the length)\n",
    "\n",
    "gunzip -c $FastQNA19238 | less -S\n",
    "\n",
    "press q to exit.\n",
    "\n",
    "    identify the read names, the sequence, the base quality scores\n",
    "\n",
    "View the reference fasta file (chr21.fa.gz) using the less command.\n",
    "\n",
    "gunzip -c $CHR21 | less -S\n",
    "\n",
    "you can scroll using the arrow keys. press q to exit.\n",
    "\n",
    "    Explain what you see\n",
    "\n",
    "Lets jump into the file on line number 200000\n",
    "\n",
    "gunzip -c $CHR21 | less -S  +200000\n",
    "\n",
    "Aligning\n",
    "\n",
    "Align the reads using bwa. We use bwa in the exercises because it is fast and widely used. We first need to index the reference chromosome, followed by the actual aligning process\n",
    "\n",
    "cp $CHR21 .\n",
    "$BWA index chr21.fa.gz\n",
    "\n",
    "    which files was generated?\n",
    "\n",
    "Lets align this single end sequencing data\n",
    "\n",
    "## align\n",
    "$BWA aln -t 5 chr21.fa.gz $FastQNA19238 > NA19238.sai\n",
    "## make a SAM file with a readGroup ID\n",
    "$BWA samse -r \"@RG\\tID:identifier\\tSM:NA19238\\tCN:BGI\\tPL:ILLUMINA\\tLB:Library\\tDS:your@mail.cn\" \\\n",
    "chr21.fa.gz NA19238.sai $FastQNA19238 | gzip -c >NA19238.sam.gz\n",
    "\n",
    "The above analysis will take couple of minuttes if you get want to can stop the analysis with CTL-C and copy the results of the commands with\n",
    "\n",
    "## if you stopped the above analysis then copy the output\n",
    "cp $COPY/NA19238.sam.gz .\n",
    "cp $COPY/NA19238.sai .\n",
    "\n",
    "    look at the generated sam file.\n",
    "\n",
    "less -S NA19238.sam.gz\n",
    "\n",
    "    Identify the header.\n",
    "    what is the content of the header\n",
    "\n",
    "    1 Find the number of reads in the fastq file.\n",
    "\n",
    "    2 Find the number of reads in the samfile.\n",
    "\n",
    "## get number of lines from fastQ file\n",
    "gunzip -c $FastQNA19238 | wc -l\n",
    "## get number of lines from the SAM file\n",
    "gunzip -c NA19238.sam.gz | wc -l\n",
    "\n",
    "Why is it not the same?\n",
    "\n",
    "Use SAMtools to sort the bam file\n",
    "\n",
    "$SAMTOOLS view -uT chr21.fa.gz NA19238.sam.gz |$SAMTOOLS sort - -o NA19238.bam\n",
    "\n",
    "BAM files faciliates random access, but this require an index this is generated using the command below.\n",
    "\n",
    "$SAMTOOLS index NA19238.bam\n",
    "\n",
    "    What is the output filename of the above command?\n",
    "\n",
    "We now have a functional alignment file that we can use for analysis.\n",
    "Viewing the data using SAMtools\n",
    "Viewing the data using tview\n",
    "\n",
    "Let try to view our aligned data\n",
    "\n",
    "$SAMTOOLS tview NA19238.bam\n",
    "\n",
    "Because we only aligned chr21 use \"g\" to go to chr21:<position> where position is the position you what to go to. NB! chr21 does not have reference bases for the first 9719760 sites.\n",
    "\n",
    "    browse the alignment using the arroy keys or space\n",
    "    look at chr21:10028371 . Why do you think so many reads mapped to this position\n",
    "    color the bases by the quality scores by pressing 'b' (you can go back pressing 'm'). press ? to see the explanation. Do you see any patterns?\n",
    "\n",
    "To get a better feel of errors and the variable sites in this genome view the alignment with the reference genome\n",
    "\n",
    "first index your fasta\n",
    "\n",
    "$SAMTOOLS faidx chr21.fa.gz\n",
    "\n",
    "and then view\n",
    "\n",
    "$SAMTOOLS tview NA19238.bam chr21.fa.gz\n",
    "\n",
    "    try to locate some variable sites e.i. sites that differ from the reference genome\n",
    "\n",
    "Vieving the data using mpileup\n",
    "\n",
    "Another way to look at the genome is by generating a pileup format\n",
    "\n",
    "$SAMTOOLS mpileup NA19238.bam | less\n",
    "\n",
    "Each line is a position with data.\n",
    "\n",
    "    When is this a usefull format?\n",
    "\n",
    "Use this command to see the depth distribution\n",
    "\n",
    "$SAMTOOLS mpileup NA19238.bam | cut -f4 | sort -n | uniq -c >dep1\n",
    "\n",
    "the left column is the number of sites and the right is the depth. View the distribution for this individuals using the following R command\n",
    "\n",
    "##open R (type R in the terminal followed by enter)\n",
    "#read in the depth and plot the first 20 depth categories\n",
    "barplot(read.table(\"dep1\")[1:20,1],names=1:20,xlab=\"depth\")\n",
    "## close R with CTRL-d or q()\n",
    "\n",
    "    if the graphics does not work you can find the plot here\n",
    "\n",
    "Remove the sequences that are likely PCR duplicates.\n",
    "\n",
    "$SAMTOOLS rmdup -s NA19238.bam NA19238.md.bam\n",
    "\n",
    "Use this command to see the depth distribution\n",
    "\n",
    "$SAMTOOLS mpileup NA19238.md.bam | cut -f4 | sort -n | uniq -c >dep2\n",
    "\n",
    "View the destribution after the duplicates are removed\n",
    "\n",
    "##open R (type R in the terminal followed by enter)\n",
    "barplot(read.table(\"dep2\")[1:20,1],names=1:20,xlab=\"depth\")\n",
    "## close R with CTRL-d or q()\n",
    "\n",
    "If the graphics does not work you can find the plot here\n",
    "\n",
    "To plot the two distributions together try\n",
    "\n",
    "##open R (type R in the terminal followed by enter)\n",
    " barplot(rbind(read.table(\"dep1\")[1:20,1],read.table(\"dep2\")[1:20,1])\n",
    ",names=1:20,xlab=\"depth\",beside=T,legend=c(\"dep1\",\"dep2\"))\n",
    "\n",
    "## close R with CTRL-d or q()\n",
    "\n",
    "if the graphics does not work you can find the plot here\n",
    "\n",
    "    Do you see a difference between the low and high depth sites?\n",
    "\n",
    "VCF file\n",
    "\n",
    "First lets view the mpileup with the reference\n",
    "\n",
    "$SAMTOOLS mpileup -f chr21.fa.gz NA19238.md.bam | less -S\n",
    "\n",
    "    What is the differnce compared to not using the reference\n",
    "\n",
    "Index the bam file\n",
    "\n",
    " $SAMTOOLS index NA19238.md.bam\n",
    "\n",
    "Lets create a VCF file for the first couple of MB of chr21. This is done based on the mpileup. There will be much more information tomorrow about how the calling is done using genotype likelihoods.\n",
    "\n",
    "$SAMTOOLS mpileup -g -f chr21.fa.gz NA19238.md.bam |  $BCFTOOL call -c -v - > sam.vcf\n",
    "\n",
    "The above analysis will take 5 of minuttes if you get want to can stop the analysis with CTL-C and copy the results of the commands with\n",
    "\n",
    "## if you stopped the above analysis then copy the output\n",
    "cp $COPY/sam.vcf .\n",
    "\n",
    "View the VCF file\n",
    "\n",
    "less -S sam.vcf\n",
    "\n",
    "    The header of the VCF contains meta information about what it in the file. In the body of the file\n",
    "\n",
    "    Identify the position, the reference allele and the alternative allele of the file.\n",
    "    Identify the depth of each position\n",
    "    Find a tri-allelic site. Do you believe that it is truely triallelic?\n",
    "    How many sites are are called as variable?\n",
    "\n",
    "Bonus exercise 1 - Paired end alignment\n",
    "\n",
    "In this exercise you will use bwa for aligning paired end reads, and try using samtools and picard tools\n",
    "\n",
    "In this exercise we have simply extracted some of the first entries in the two fastq files, then we aligned it and generated a SAMfile using the commands shown below. The command for aligment of the paired end is below however, it takes too long to run. Instead you can view the output with the name $SAMFILE\n",
    "\n",
    "Aligning the data\n",
    "\n",
    "### dont run example in this block, takes to long due to the full genome reference\n",
    "# $BWA aln -I -t 10 $HG19 fq/id_1.fq.gz >id_1.sai\n",
    "# $BWA aln -I -t 10 $HG19 fq/id_2.fq.gz >id_2.sai\n",
    "#$BWA sampe -P $HG19 id_1.sai id_2.sai  fq/id_1.fq.gz fq/id_2.fq.gz \\\n",
    "-r  \"@RG\\tID:myRgId\\tSM:mySample\\tCN:BGI\\tPL:ILLUMINA\\tLB:myLib\\tDS:my@mail.cn\" \\\n",
    "# |gzip -c >id.sam.gz\n",
    "\n",
    "Look at the sam file ($SAMfile) and identify the 10 different columns (samtools manual)\n",
    "\n",
    "less -S $SAMfile\n",
    "\n",
    "What is the distribution of reads per \"chromosome\" based on the RNAME column ? You can use use the 'cut', 'sort', and 'uniq' UNIX commands\n",
    "\n",
    "gunzip -c $SAMfile | grep -P \"^@\" -v |cut -f3|sort |uniq -c\n",
    "\n",
    "Using Picardtools\n",
    "\n",
    "bwa actually fills in the mate information, but not all aligners do that, so we can run picard tools to fill in the mate information and sort the file according to position. We will output the file in the binary version of SAM which is BAM\n",
    "\n",
    "java -jar $PICARD FixMateInformation INPUT=$SAMfile \\\n",
    "OUTPUT=id.fixmate.srt.bam SORT_ORDER=coordinate\n",
    "\n",
    "View the header of the BAM file\n",
    "\n",
    "$SAMTOOLS view -H id.fixmate.srt.bam\n",
    "\n",
    "picard didn't update the PG flag, so let us update the header information so that we have documented how we modified the file.\n",
    "\n",
    "($SAMTOOLS view -H id.fixmate.srt.bam;echo -e \"@PG\\tID:fixmate\\tPN:fixmate\\tVN:2.60\\tCL:stuff\" ) \\\n",
    ">newhd\n",
    "$SAMTOOLS reheader newhd id.fixmate.srt.bam > id.fixmate.srt2.bam\n",
    "\n",
    "    Validate that the header in file id.fixmate.srt2.bam has been updated\n",
    "\n",
    "Now mark duplicates using picard\n",
    "\n",
    "java -jar $PICARD MarkDuplicates I=id.fixmate.srt2.bam \\\n",
    "O=id.fixmate.srt.md.bam  M=metrics;\n",
    "\n",
    "    Did picard update the PG flag of the header?\n",
    "    Did picard update anything else in the header?\n",
    "\n",
    "NB you can view the header of a bamfile using 'samtools view -H'\n",
    "Bonus exercise 2 - clean you bam files using the FLAGS column\n",
    "\n",
    "The second column in the SAM format is the very important FLAG. This will tell tell you about the state of the paired end mapping, QC duplicates etc.\n",
    "\n",
    "When we worked with the paired end mapping using picard earlier we had the following temporary files.\n",
    "\n",
    "    id.sam.gz ($SAMfile)\n",
    "    id.fixmate.srt2.bam\n",
    "    id.fixmate.srt.md.bam\n",
    "\n",
    "Using the samtools -F/-f you can discard/include flags that fulfill certain patterns. See http://broadinstitute.github.io/picard/explain-flags.html .\n",
    "\n",
    "    How many reads have we marked as duplicate in the final file.\n",
    "    How many properly mapped read pairs do we have? (Where both reads map to the same chr etc).\n",
    "    How many mapped reads do we have ?\n",
    "    How many unmapped reads do we have ?\n",
    "    Find the distribution of the RNAMES of the unmapped reads!?\n",
    "    Make a new bamfile, where you only the reads where both ends maps, and filter out those with a mapping quality below 10, and removing duplicates\n",
    "\n",
    "Run the following command one at a time\n",
    "\n",
    "$SAMTOOLS view -f 1024 id.fixmate.srt.md.bam|wc -l\n",
    "$SAMTOOLS view -f 2 id.fixmate.srt.md.bam|wc -l\n",
    "$SAMTOOLS view -F 4 id.fixmate.srt.md.bam|wc -l\n",
    "$SAMTOOLS view -f 4 id.fixmate.srt.md.bam|wc -l\n",
    "$SAMTOOLS view -f 4 id.fixmate.srt.md.bam|cut -f3|sort -n |uniq -c\n",
    "$SAMTOOLS view -f 2 -F 1024 id.fixmate.srt.md.bam -q 10 >new.bam\n",
    "\n",
    "    Compare with \"samtools flagstat\" command\n",
    "\n",
    "$SAMTOOLS flagstat id.fixmate.srt.md.bam\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After loggin into the server then paste the follow into your terminal to set paths\n",
    "\n",
    "DATA=/ricco/data/PhDCourse/summer2019/ngs/\n",
    "SAMfile=$DATA/id.sam.gz\n",
    "FastQNA19238=$DATA/NA19238.fastq.gz\n",
    "CHR21=$DATA/chr21.fa.gz\n",
    "HG19=$DATA/hg19.fa.gz\n",
    "COPY=$DATA/preComp/\n",
    "\n",
    "# programs\n",
    "SAMTOOLS=samtools\n",
    "BWA=bwa\n",
    "PICARD=$DATA/picard.jar\n",
    "ANGSD=angsd\n",
    "BCFTOOL=bcftools\n",
    "\n",
    "Mapping one reduced genome (single end)\n",
    "\n",
    "In this exercise you will align a fastq file using bwa and generate a SAM file.\n",
    "\n",
    "Due to the computational time we have created a reduced genome from one of the individuals from the 1000genomes pilot project. The individual, NA19238, has been sequenced using solexa/Illumina. For this excerice we have created a reduced reference genome consisting only of chromosome 21 (the smallest human chromosome) and also reduced the sequencing data to reads that will likely map to chromosome 21.\n",
    "\n",
    "    The fastQ file 'NA19238.fastq.gz' has variable name $FastQNA19238 which is the sample we want to align and the chr21.fa.gz ($CHR21) reference genome that we want to align against.\n",
    "\n",
    "Viewing the input files\n",
    "\n",
    "    view the fastq file (NA19238.fastq) using the less command and identity the reads and quality scores (ignore the length)\n",
    "\n",
    "gunzip -c $FastQNA19238 | less -S\n",
    "\n",
    "press q to exit.\n",
    "\n",
    "    identify the read names, the sequence, the base quality scores\n",
    "\n",
    "View the reference fasta file (chr21.fa.gz) using the less command.\n",
    "\n",
    "gunzip -c $CHR21 | less -S\n",
    "\n",
    "you can scroll using the arrow keys. press q to exit.\n",
    "\n",
    "    Explain what you see\n",
    "\n",
    "Lets jump into the file on line number 200000\n",
    "\n",
    "gunzip -c $CHR21 | less -S  +200000\n",
    "\n",
    "Aligning\n",
    "\n",
    "Align the reads using bwa. We use bwa in the exercises because it is fast and widely used. We first need to index the reference chromosome, followed by the actual aligning process\n",
    "\n",
    "cp $CHR21 .\n",
    "$BWA index chr21.fa.gz\n",
    "\n",
    "    which files was generated?\n",
    "\n",
    "Lets align this single end sequencing data\n",
    "\n",
    "## align\n",
    "$BWA aln -t 5 chr21.fa.gz $FastQNA19238 > NA19238.sai\n",
    "## make a SAM file with a readGroup ID\n",
    "$BWA samse -r \"@RG\\tID:identifier\\tSM:NA19238\\tCN:BGI\\tPL:ILLUMINA\\tLB:Library\\tDS:your@mail.cn\" \\\n",
    "chr21.fa.gz NA19238.sai $FastQNA19238 | gzip -c >NA19238.sam.gz\n",
    "\n",
    "The above analysis will take couple of minuttes if you get want to can stop the analysis with CTL-C and copy the results of the commands with\n",
    "\n",
    "## if you stopped the above analysis then copy the output\n",
    "cp $COPY/NA19238.sam.gz .\n",
    "cp $COPY/NA19238.sai .\n",
    "\n",
    "    look at the generated sam file.\n",
    "\n",
    "less -S NA19238.sam.gz\n",
    "\n",
    "    Identify the header.\n",
    "    what is the content of the header\n",
    "\n",
    "    1 Find the number of reads in the fastq file.\n",
    "\n",
    "    2 Find the number of reads in the samfile.\n",
    "\n",
    "## get number of lines from fastQ file\n",
    "gunzip -c $FastQNA19238 | wc -l\n",
    "## get number of lines from the SAM file\n",
    "gunzip -c NA19238.sam.gz | wc -l\n",
    "\n",
    "Why is it not the same?\n",
    "\n",
    "Use SAMtools to sort the bam file\n",
    "\n",
    "$SAMTOOLS view -uT chr21.fa.gz NA19238.sam.gz |$SAMTOOLS sort - -o NA19238.bam\n",
    "\n",
    "BAM files faciliates random access, but this require an index this is generated using the command below.\n",
    "\n",
    "$SAMTOOLS index NA19238.bam\n",
    "\n",
    "    What is the output filename of the above command?\n",
    "\n",
    "We now have a functional alignment file that we can use for analysis.\n",
    "Viewing the data using SAMtools\n",
    "Viewing the data using tview\n",
    "\n",
    "Let try to view our aligned data\n",
    "\n",
    "$SAMTOOLS tview NA19238.bam\n",
    "\n",
    "Because we only aligned chr21 use \"g\" to go to chr21:<position> where position is the position you what to go to. NB! chr21 does not have reference bases for the first 9719760 sites.\n",
    "\n",
    "    browse the alignment using the arroy keys or space\n",
    "    look at chr21:10028371 . Why do you think so many reads mapped to this position\n",
    "    color the bases by the quality scores by pressing 'b' (you can go back pressing 'm'). press ? to see the explanation. Do you see any patterns?\n",
    "\n",
    "To get a better feel of errors and the variable sites in this genome view the alignment with the reference genome\n",
    "\n",
    "first index your fasta\n",
    "\n",
    "$SAMTOOLS faidx chr21.fa.gz\n",
    "\n",
    "and then view\n",
    "\n",
    "$SAMTOOLS tview NA19238.bam chr21.fa.gz\n",
    "\n",
    "    try to locate some variable sites e.i. sites that differ from the reference genome\n",
    "\n",
    "Vieving the data using mpileup\n",
    "\n",
    "Another way to look at the genome is by generating a pileup format\n",
    "\n",
    "$SAMTOOLS mpileup NA19238.bam | less\n",
    "\n",
    "Each line is a position with data.\n",
    "\n",
    "    When is this a usefull format?\n",
    "\n",
    "Use this command to see the depth distribution\n",
    "\n",
    "$SAMTOOLS mpileup NA19238.bam | cut -f4 | sort -n | uniq -c >dep1\n",
    "\n",
    "the left column is the number of sites and the right is the depth. View the distribution for this individuals using the following R command\n",
    "\n",
    "##open R (type R in the terminal followed by enter)\n",
    "#read in the depth and plot the first 20 depth categories\n",
    "barplot(read.table(\"dep1\")[1:20,1],names=1:20,xlab=\"depth\")\n",
    "## close R with CTRL-d or q()\n",
    "\n",
    "    if the graphics does not work you can find the plot here\n",
    "\n",
    "Remove the sequences that are likely PCR duplicates.\n",
    "\n",
    "$SAMTOOLS rmdup -s NA19238.bam NA19238.md.bam\n",
    "\n",
    "Use this command to see the depth distribution\n",
    "\n",
    "$SAMTOOLS mpileup NA19238.md.bam | cut -f4 | sort -n | uniq -c >dep2\n",
    "\n",
    "View the destribution after the duplicates are removed\n",
    "\n",
    "##open R (type R in the terminal followed by enter)\n",
    "barplot(read.table(\"dep2\")[1:20,1],names=1:20,xlab=\"depth\")\n",
    "## close R with CTRL-d or q()\n",
    "\n",
    "If the graphics does not work you can find the plot here\n",
    "\n",
    "To plot the two distributions together try\n",
    "\n",
    "##open R (type R in the terminal followed by enter)\n",
    " barplot(rbind(read.table(\"dep1\")[1:20,1],read.table(\"dep2\")[1:20,1])\n",
    ",names=1:20,xlab=\"depth\",beside=T,legend=c(\"dep1\",\"dep2\"))\n",
    "\n",
    "## close R with CTRL-d or q()\n",
    "\n",
    "if the graphics does not work you can find the plot here\n",
    "\n",
    "    Do you see a difference between the low and high depth sites?\n",
    "\n",
    "VCF file\n",
    "\n",
    "First lets view the mpileup with the reference\n",
    "\n",
    "$SAMTOOLS mpileup -f chr21.fa.gz NA19238.md.bam | less -S\n",
    "\n",
    "    What is the differnce compared to not using the reference\n",
    "\n",
    "Index the bam file\n",
    "\n",
    " $SAMTOOLS index NA19238.md.bam\n",
    "\n",
    "Lets create a VCF file for the first couple of MB of chr21. This is done based on the mpileup. There will be much more information tomorrow about how the calling is done using genotype likelihoods.\n",
    "\n",
    "$SAMTOOLS mpileup -g -f chr21.fa.gz NA19238.md.bam |  $BCFTOOL call -c -v - > sam.vcf\n",
    "\n",
    "The above analysis will take 5 of minuttes if you get want to can stop the analysis with CTL-C and copy the results of the commands with\n",
    "\n",
    "## if you stopped the above analysis then copy the output\n",
    "cp $COPY/sam.vcf .\n",
    "\n",
    "View the VCF file\n",
    "\n",
    "less -S sam.vcf\n",
    "\n",
    "    The header of the VCF contains meta information about what it in the file. In the body of the file\n",
    "\n",
    "    Identify the position, the reference allele and the alternative allele of the file.\n",
    "    Identify the depth of each position\n",
    "    Find a tri-allelic site. Do you believe that it is truely triallelic?\n",
    "    How many sites are are called as variable?\n",
    "\n",
    "Bonus exercise 1 - Paired end alignment\n",
    "\n",
    "In this exercise you will use bwa for aligning paired end reads, and try using samtools and picard tools\n",
    "\n",
    "In this exercise we have simply extracted some of the first entries in the two fastq files, then we aligned it and generated a SAMfile using the commands shown below. The command for aligment of the paired end is below however, it takes too long to run. Instead you can view the output with the name $SAMFILE\n",
    "\n",
    "Aligning the data\n",
    "\n",
    "### dont run example in this block, takes to long due to the full genome reference\n",
    "# $BWA aln -I -t 10 $HG19 fq/id_1.fq.gz >id_1.sai\n",
    "# $BWA aln -I -t 10 $HG19 fq/id_2.fq.gz >id_2.sai\n",
    "#$BWA sampe -P $HG19 id_1.sai id_2.sai  fq/id_1.fq.gz fq/id_2.fq.gz \\\n",
    "-r  \"@RG\\tID:myRgId\\tSM:mySample\\tCN:BGI\\tPL:ILLUMINA\\tLB:myLib\\tDS:my@mail.cn\" \\\n",
    "# |gzip -c >id.sam.gz\n",
    "\n",
    "Look at the sam file ($SAMfile) and identify the 10 different columns (samtools manual)\n",
    "\n",
    "less -S $SAMfile\n",
    "\n",
    "What is the distribution of reads per \"chromosome\" based on the RNAME column ? You can use use the 'cut', 'sort', and 'uniq' UNIX commands\n",
    "\n",
    "gunzip -c $SAMfile | grep -P \"^@\" -v |cut -f3|sort |uniq -c\n",
    "\n",
    "Using Picardtools\n",
    "\n",
    "bwa actually fills in the mate information, but not all aligners do that, so we can run picard tools to fill in the mate information and sort the file according to position. We will output the file in the binary version of SAM which is BAM\n",
    "\n",
    "java -jar $PICARD FixMateInformation INPUT=$SAMfile \\\n",
    "OUTPUT=id.fixmate.srt.bam SORT_ORDER=coordinate\n",
    "\n",
    "View the header of the BAM file\n",
    "\n",
    "$SAMTOOLS view -H id.fixmate.srt.bam\n",
    "\n",
    "picard didn't update the PG flag, so let us update the header information so that we have documented how we modified the file.\n",
    "\n",
    "($SAMTOOLS view -H id.fixmate.srt.bam;echo -e \"@PG\\tID:fixmate\\tPN:fixmate\\tVN:2.60\\tCL:stuff\" ) \\\n",
    ">newhd\n",
    "$SAMTOOLS reheader newhd id.fixmate.srt.bam > id.fixmate.srt2.bam\n",
    "\n",
    "    Validate that the header in file id.fixmate.srt2.bam has been updated\n",
    "\n",
    "Now mark duplicates using picard\n",
    "\n",
    "java -jar $PICARD MarkDuplicates I=id.fixmate.srt2.bam \\\n",
    "O=id.fixmate.srt.md.bam  M=metrics;\n",
    "\n",
    "    Did picard update the PG flag of the header?\n",
    "    Did picard update anything else in the header?\n",
    "\n",
    "NB you can view the header of a bamfile using 'samtools view -H'\n",
    "Bonus exercise 2 - clean you bam files using the FLAGS column\n",
    "\n",
    "The second column in the SAM format is the very important FLAG. This will tell tell you about the state of the paired end mapping, QC duplicates etc.\n",
    "\n",
    "When we worked with the paired end mapping using picard earlier we had the following temporary files.\n",
    "\n",
    "    id.sam.gz ($SAMfile)\n",
    "    id.fixmate.srt2.bam\n",
    "    id.fixmate.srt.md.bam\n",
    "\n",
    "Using the samtools -F/-f you can discard/include flags that fulfill certain patterns. See http://broadinstitute.github.io/picard/explain-flags.html .\n",
    "\n",
    "    How many reads have we marked as duplicate in the final file.\n",
    "    How many properly mapped read pairs do we have? (Where both reads map to the same chr etc).\n",
    "    How many mapped reads do we have ?\n",
    "    How many unmapped reads do we have ?\n",
    "    Find the distribution of the RNAMES of the unmapped reads!?\n",
    "    Make a new bamfile, where you only the reads where both ends maps, and filter out those with a mapping quality below 10, and removing duplicates\n",
    "\n",
    "Run the following command one at a time\n",
    "\n",
    "$SAMTOOLS view -f 1024 id.fixmate.srt.md.bam|wc -l\n",
    "$SAMTOOLS view -f 2 id.fixmate.srt.md.bam|wc -l\n",
    "$SAMTOOLS view -F 4 id.fixmate.srt.md.bam|wc -l\n",
    "$SAMTOOLS view -f 4 id.fixmate.srt.md.bam|wc -l\n",
    "$SAMTOOLS view -f 4 id.fixmate.srt.md.bam|cut -f3|sort -n |uniq -c\n",
    "$SAMTOOLS view -f 2 -F 1024 id.fixmate.srt.md.bam -q 10 >new.bam\n",
    "\n",
    "    Compare with \"samtools flagstat\" command\n",
    "\n",
    "$SAMTOOLS flagstat id.fixmate.srt.md.bam\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
