\documentclass[11pt]{article}
\title {MRes Proposal: Spatial genomics of a ubiquitous and devastating bumble bee pathogen (Nosema bombi).}
\author{Victoria Blanchard}
\date{2 Dec 2019}
\usepackage[margin=2cm]{geometry}
\usepackage{graphicx}
\usepackage[]{lineno}
\usepackage{booktabs}
\usepackage[backend=biber,style=authoryear,bibencoding=utf8]{biblatex}
\setlength{\parindent}{0em}


% bibliography secion %

\addbibresource{MResReferences.bib}
\newcommand{\ra}[1]{\renewcommand{\arraystretch}{#1}}

\begin{document}

\begin{titlepage}


	\centering % this centers everything on the page
		
	%\vspace  % Whitespace at the top of the page 
	
	
	% --------------------------
	%% TITLE
	
	\vspace*{5\baselineskip}
	
	\rule{\textwidth}{1.6pt}\vspace*{-\baselineskip}\vspace*{2pt} % Thick horizontal rule
	\rule{\textwidth}{0.4pt} % Thin horizontal rule
	
	\vspace{0.75\baselineskip} % Whitespace above the title
	
	{\LARGE SPATIAL GENOMICS OF A \\ UBIQUITOUS AND DEVASTATING \\ BUMBLE BEE PATHOGEN (\textit{Nosema bombi})\\} 
	
	\vspace{0.75\baselineskip} % Whitespace below the title
	
	\rule{\textwidth}{0.4pt}\vspace*{-\baselineskip}\vspace{3.2pt} 
	\rule{\textwidth}{1.6pt} 
	
	\vspace{2\baselineskip} 
	
	% ---------------------------------
	%% SUPERVISORS & CONTACT EMAIL
	
	Supervisors: \\
		Professor Mark Brown, Royal Holloway, University of London \\
		Professor  Matthew Fisher, Imperial College London 

		
	\vspace{1.5 \baselineskip} % Whitespace between text
	
	Contact: \\
	vlb19@imperial.ac.uk
	


\end{titlepage}

\linenumbers

	\section{Key Words}
	\begin{itemize}
	
	\item Microsporidian
	\item \textit{Bombus}
	\item Pathogen
	\item Genetics
	
	\end{itemize}
	
	\section{Introduction}

Bumblebees provide an important pollination service globally (Cameron et al.). European bumblebees are often infected with a microsporidian parasite (Nosema bombi) (Brown 2017) which has been suggested to have been introduced to North America in the 1990s, and be the causal factor behind massive rapid declines in North American bumblebees (Thorp et al 2008 , Cameron et al. 2011, Cameron  et al. 2016). All North American declining bumblebee populations are infected with nosema – with prevalances between 15 and 37\% within a single species (Cameron  et al. 2016). However, the most recent study into genetic differences between North American and European Nosema bombi populations failed to detect variation between the two regions (Cameron et al. 2010). The absence of a complete genome may lie behind these results, as previous studies have used less complete genetic resources which may lack power to identify relevant variation (Cameron et al. 2016). Here we propose to assemble a genome for N. bombi that enables high-powered spatial and temporal population genomics. To test its efficacy, we will assess how land-use patterns link to population genomics structure in this pathogen.	

	\section{Questions and project goals}
	
\begin{itemize}
	\item Create a de novo genome for \textit{N. bombi}
	\item Create a barcoded library of a set of samples
	\item Investigate spatial variation in population genomics of \textit{N. bombi} throughout the mainland UK
	\item Compare UK samples with US MLST data
	\item Assess the impact of land use on genomic variability in UK samples
\end{itemize}

	\section{Methods}

\begin{enumerate}

	\item \textit{Sample collection} \\

Retrieve samples from storage at -80oC. Transfer 15ul of inoculum onto a haemocytometer and count number of nosema spores using established lab protocol (Folly and Brown DATE). Extract genomic DNA using the appropriate kit – selected closer to the time. Amplify DNA using custom primers with a negative water control. Repeat each PCR in triplicate and visualise on agarose gel. Purify DNA using the appropriate kit then quantify amount of DNA (Bates et al. 2018).

	\item \textit{Assembling genome} \\

Use a MiSeq instrument to quantify number of reads yielded per sample from the preliminary library. Create a final composite library based on the index representation from the initial MiSeq run, then run a final sequence using the MiSeq with the appropriate paired-end strategy. Analyse and quality-filter the sequences using MOTHUR. Cluster rRNA gene sequences into groups according to taxonomy. Eliminate sequences derived from chloroplasts, mitochondria, archara, eukaryotes, and unknown reads. Analyse OTUs using the Phyloseq package in R (Bates et al. 2018).

	\item \textit{Creating barcode library} \\

Create a library of short DNA sequences which can identify the organism to species. We will test whether the COX1 gene is an appropriate choice for this, or use internal transcribed spacer (ITS) rNA. There are candidate markers identified in Cameron et al. (2016) which we will also use as fungal barcoding may require more than one primer combination.  This can then be used by any other researchers to identify the species of their microsporidian.

	\item \textit{Comparing population genomics} \\

To search for differences between populations we will use RepeatModeler to identify new repeats from the assemblies. This will allow us to characterise any novel species should they arise (Farrer et al. 2017). We will then do sequence alignments using HPC to identify base pair differences.

	\item \textit{Assessing impacts of land use} \\

Analyse infection intensity between different samples and use a GLM to establish if intensity differs with year or geographic location. We can characterise land use of sampled area using GIS mapping within the average foraging distance of a bumblebee - a 1.5km radius of the capture site (Osborne et al. 2008), and include this as a covariate in the model.

\end{enumerate}
	
\begin{figure}[h!]
	
	\includegraphics[width = \linewidth]{Gantt_Image.png}
	\caption{	Gantt Chart}
	\label{MRes Gantt}
	
\end{figure} 

\newpage
	\section{Budget}

	\begin{table*}[h]\centering
	\ra{1.3}
		\begin{tabular} {@{}rcrcrc@{}}\toprule
		\hline
		\textbf{Item} & \textbf{Cost} & \textbf{Justification} \\
		\hline\hline
		MiSeq & \pounds1000 & Needed for assembling the \\
		& & genome from microsporidian \\
		\hline
		Commuting to South Kensington & \pounds210 &	Travelling to Imperial College (South Kensington campus) \\
		& & 5 days a week for 4 weeks to use facilities \\
		\hline
		Travelling to field sites & \pounds300 & Fuel costs for travelling to field sites for data collection\\
		\hline
		\bottomrule
		\end{tabular}
	\caption{Detailed budget break-down}
	\end{table*}

	  
\newpage

	\section{References}
A field experiment on the effect of Nosema bombi in colonies of the bumblebee Bombus terrestris. Otti and Schmid-Hempel \\

Amphibian chytridiomycosis outbreak dynamics are linked with host skin bacterial community structure Bates et al. 2018 \\ 

Genomic innovations linked to infection strategies across emerging pathogenic chytrid fungi Farrer et al. 2017 \\

Patterns of widespread decline in North American bumble bees. Cameron SA1, Lozier JD, Strange JP, Koch JB, Cordes N, Solter LF, Griswold TL \\

Colony success of the bumble bee, Bombus terrestris, in relation to infections by two protozoan parasites, Crithidia bombi and Nosema bombi . Imhoof and P. Schmid-Hempel \\

Within colony dynamics of Nosema bombi infections: disease establishment, epidemiology and potential vertical transmission. Samina T. Rutrecht1,2, Mark J.F. Brown \\

A contribution to the knowledge of Nosema infections in bumble bees, Bombus spp. Paul Schmid-Hempel Roland Loosli\\

Test of the invasive pathogen hypothesis of bumble bee decline in North America Sydney A. Cameron,a,1 Haw Chuan Lim,a,2 Jeffrey D. Lozier,b Michelle A. Duennes,a,3 and Robbin Thorp \\

MiSeq: A Next Generation Sequencing Platform for Genomic Analysis Rupesh Kanchi Ravi, Kendra Walton, Mahdieh Khosroheidaris \\

Landscape predictors of pathogen prevalence and range contractions in US bumblebees Scott H. McArt1 , Christine Urbanowicz2 , Shaun McCoshum3 , Rebecca E. Irwin4 and Lynn S. Adler \\

Bumblebee flight distances in relation to the forage landscape. Osborne JL1, Martin AP, Carreck NL, Swain JL, Knight ME, Goulson D, Hale RJ, Sanderson RA. \\

\newpage

	\section{Supervisor approval}

	
	I have seen and approved the proposal and the budget. \\
	
	
	Supervisor name: Professor Mark Brown \\
	
	 
	
	Supervisor signature: 

\begin{figure}[h!]
	
	\includegraphics[width = 6cm]{MarkSignature.png}
	
\end{figure} 

\ref{eq:quad}\

\ref{eq:cubic}\

\end{document}